File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/Activity/EditUserActivity.java
package com.ducluanxutrieu.quanlynhanvien.Activity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.ducluanxutrieu.quanlynhanvien.Models.Users;
import com.ducluanxutrieu.quanlynhanvien.R;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.functions.FirebaseFunctions;
import com.google.firebase.functions.FirebaseFunctionsException;
import com.google.firebase.functions.HttpsCallableResult;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.StorageTask;
import com.google.firebase.storage.UploadTask;

import java.util.HashMap;

public class EditUserActivity extends AppCompatActivity {
    TextInputEditText name, email, phone, position, password;
    LinearLayout inputPasswordLayout;
    Button add, cancel;
    String signal;
    boolean isAdmin;
    Users users;
    ImageView avatar;

    String mName = "";
    static String mEmail = "";
    static String mPhone = "";
    static String mPosition = "";
    static String mPassword = "";

    StorageReference mStorageReference;
    DatabaseReference mDatabaseReference;
    FirebaseFunctions mFunctions;

    private static final int IMAGE_REQUEST = 1;
    private static final String TAG = "kiemtra";
    private Uri uriImage;
    private StorageTask uploadTask;

    ProgressDialog pd;


    private String getStr(EditText editText){
        return editText.getText().toString();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_user);

        //create progressDialog
        pd = new ProgressDialog(EditUserActivity.this);
        mFunctions = FirebaseFunctions.getInstance();

        //get Data and signal from parent activity
        Intent intent = getIntent();
        users = (Users) intent.getSerializableExtra("user");
        signal = intent.getStringExtra("signal");
        isAdmin = intent.getBooleanExtra("isAdmin", false);

        mapping();

        if (signal.equals("updateUser")){
            avatar.setEnabled(true);
            mStorageReference = FirebaseStorage.getInstance().getReference().child("Users/" + users.getUid() + "/avatar");
            if (users.getName() != null) {
                name.setText(users.getName());
            }
            if (users.getEmail() != null) {
                email.setText(users.getEmail());
            }
            if (users.getPhone() != null) {
                phone.setText(users.getPhone());
            }
            if (users.getPosition() != null) {
                position.setText(users.getPosition());
            }
            if (users.getPassword() != null){
                password.setText(users.getPassword());
            }
            add.setText(getString(R.string.update));
            if (users.getAvatarUrl() != null) {
                Glide.with(this).load(users.getAvatarUrl()).apply(RequestOptions.circleCropTransform()).into(avatar);
            }
        }else {
            avatar.setEnabled(false);
        }
        add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                pd.setTitle("Uploading...");
                pd.show();
                mName = getStr(name);
                //mName =( name.getText().toString() == null ? "" : name.getText().toString());
                mEmail = getStr(email);
                mPhone = getStr(phone);
                mPosition = getStr(position);
                mPassword = getStr(password);

                if (mPhone.startsWith("0")){
                    mPhone = "+84" + mPhone.substring(1);
                }

                HashMap<String, Object> map;
                if (putToHashMap() == null){
                    return;
                }else {
                    map = putToHashMap();
                }

                modifyUser(map, signal).addOnCompleteListener(new OnCompleteListener<String>() {
                    @Override
                    public void onComplete(@NonNull Task<String> task) {
                        if (!task.isSuccessful()) {
                            Exception e = task.getException();
                            if (e != null) {
                                Log.i(TAG, e.toString());
                            }
                            if (e instanceof FirebaseFunctionsException) {
                                FirebaseFunctionsException ffe = (FirebaseFunctionsException) e;
                                FirebaseFunctionsException.Code code = ffe.getCode();
                                //Object details = ffe.getDetails();
                                //Log.i(TAG, details.toString());
                                Log.i(TAG, code.toString());
                            }
                            // [START_EXCLUDE]
                            Log.w(TAG, "addNumbers:onFailure", e);
                            showSnackbar("An error occurred.");
                            // [END_EXCLUDE]
                            // ...
                        }else {
                            // [START_EXCLUDE]
                            String result = task.getResult();

                            // [END_EXCLUDE]
                            Toast.makeText(EditUserActivity.this, result, Toast.LENGTH_SHORT).show();
                        }

                        // ...
                    }

                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Toast.makeText(EditUserActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });

                if (signal.equals("updateUser")) {
                    uploadImage();
                }
                pd.dismiss();
                finish();
            }
        });

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        avatar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openImage();
            }
        });
    }

    private HashMap<String, Object> putToHashMap() {
        HashMap<String, Object> map = new HashMap<>();
        if (mName == null || mName.isEmpty()){
            return null;
        }else {
            map.put("name", mName);
        }

        if (mEmail == null || mEmail.isEmpty()){
            return null;
        }else {
            map.put("email", mEmail);
        }

        map.put("phone", mPhone);
        map.put("position", mPosition);

        if (mPassword == null || mPassword.isEmpty()){
            return null;
        }else {
            map.put("password", mPassword);
        }
        if (users != null){
            map.put("uid", users.getUid());
            map.put("avatarUrl", users.getAvatarUrl());
        }
        return map;
    }

    private void openImage() {
        Intent intent = new Intent();
        intent.setType("image/*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(intent, IMAGE_REQUEST);
    }

/*    private String getFileExtension(Uri uri){
        ContentResolver contentResolver = this.getContentResolver();
        MimeTypeMap mimeTypeMap = MimeTypeMap.getSingleton();
        return mimeTypeMap.getExtensionFromMimeType(contentResolver.getType(uri));
    }*/

    private void uploadImage(){
        if (uriImage!= null){
            //final StorageReference storageReference = mStorageReference.child(System.currentTimeMillis() + "." + getFileExtension(uriImage));

            uploadTask = mStorageReference.putFile(uriImage);
            uploadTask.continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
                @Override
                public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
                    if (!task.isSuccessful()){
                        throw task.getException();
                    }
                    return mStorageReference.getDownloadUrl();
                }
            }).addOnCompleteListener(new OnCompleteListener() {
                @Override
                public void onComplete(@NonNull Task task) {
                    if (task.isSuccessful()){
                        Uri downloadUri = (Uri) task.getResult();
                        if(downloadUri != null) {
                            String mUri = downloadUri.toString();
                            mDatabaseReference = FirebaseDatabase.getInstance().getReference().child("users/" + users.getUid());
                            HashMap<String, Object> map = new HashMap<>();
                            map.put("avatarUrl", mUri);
                            mDatabaseReference.updateChildren(map);
                        }else {
                            Toast.makeText(EditUserActivity.this, getString(R.string.can_not_upload_image), Toast.LENGTH_SHORT).show();
                        }
                    }
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    Toast.makeText(EditUserActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                    Log.i(TAG, e.getMessage());
                }
            });
        }else {
            Toast.makeText(this, "No image selected", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == IMAGE_REQUEST && resultCode == RESULT_OK && data!= null && data.getData() != null){
            uriImage = data.getData();
            if (uploadTask != null && uploadTask.isInProgress()){
                Toast.makeText(this, "Upload in progress", Toast.LENGTH_SHORT).show();
            }else {
                Glide.with(this).load(uriImage).apply(RequestOptions.circleCropTransform()).into(avatar);
            }
        }
    }

    public void mapping() {
        name = findViewById(R.id.edit_text_name);
        email = findViewById(R.id.edit_text_email);
        phone = findViewById(R.id.edit_text_phone);
        position = findViewById(R.id.edit_text_position);
        add = findViewById(R.id.btn_add);
        cancel = findViewById(R.id.btn_cancel);
        avatar = findViewById(R.id.user_profile_photo_edit);
        password = findViewById(R.id.edit_text_password);
        inputPasswordLayout = findViewById(R.id.edit_text_password_layout);
    }

    private Task<String> modifyUser(HashMap<String, Object> data, String signal) {
        // Create the arguments to the callable function.
        Log.i(TAG, signal);
        return mFunctions
                .getHttpsCallable(signal)
                .call(data)
                .continueWith(new Continuation<HttpsCallableResult, String>() {
                    @Override
                    public String then(@NonNull Task<HttpsCallableResult> task) throws Exception {
                        // This continuation runs on either success or failure, but if the task
                        // has failed then getResult() will throw an Exception which will be
                        // propagated down.
                        return (String) task.getResult().getData();
                    }
                });
    }

    private void showSnackbar(String message) {
        Snackbar.make(findViewById(android.R.id.content), message, Snackbar.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/activity/EditUserActivity.kt
package com.ducluanxutrieu.quanlynhanvien.activity

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.ducluanxutrieu.quanlynhanvien.R
import com.ducluanxutrieu.quanlynhanvien.activity.MainActivity.Companion.mCatView
import com.ducluanxutrieu.quanlynhanvien.models.Staff
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.functions.FirebaseFunctionsException
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.google.firebase.storage.UploadTask
import java.util.*

class EditUserActivity : AppCompatActivity() {
    internal lateinit var name: TextInputEditText
    internal lateinit var email: TextInputEditText
    private lateinit var phone: TextInputEditText
    internal lateinit var position: TextInputEditText
    private lateinit var password: TextInputEditText
    private lateinit var inputPasswordLayout: LinearLayout
    internal lateinit var add: Button
    internal lateinit var cancel: Button
    private var signal: String? = null
    private var isAdmin: Boolean = false
    private var staff: Staff? = null
    private lateinit var avatar: ImageView

    private var mName: String? = ""

    private lateinit var mStorageReference: StorageReference
    private lateinit var mDatabaseReference: DatabaseReference
    private lateinit var mFunctions: FirebaseFunctions
    private var uriImage: Uri? = null
    private var uploadTask: UploadTask? = null

    private fun getStr(editText: EditText): String {
        return editText.text.toString()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_user)

        mFunctions = FirebaseFunctions.getInstance()

        //get Data and signal from parent activity
        val intent = intent
        staff = intent.getSerializableExtra("user") as Staff
        signal = intent.getStringExtra("signal")
        isAdmin = intent.getBooleanExtra("isAdmin", false)

        binding()

        updateUser()

        cancel.setOnClickListener { finish() }

        avatar.setOnClickListener { openImage() }
    }

    private fun updateUser() {
        if (signal == "updateUser") {
            avatar.isEnabled = true
            mStorageReference = FirebaseStorage.getInstance().reference.child("Staff/" + staff!!.uid + "/avatar")
            if (staff!!.name != null) {
                name.setText(staff!!.name)
            }
            if (staff!!.email != null) {
                email.setText(staff!!.email)
            }
            if (staff!!.phone != null) {
                phone.setText(staff!!.phone)
            }
            if (staff!!.position != null) {
                position.setText(staff!!.position)
            }
            if (staff!!.password != null) {
                password.setText(staff!!.password)
            }
            add.text = getString(R.string.update)
            if (staff!!.avatarUrl != null) {
                Glide.with(this).load(staff!!.avatarUrl).apply(RequestOptions.circleCropTransform()).into(avatar)
            }
        } else {
            avatar.isEnabled = false
        }
    }

    fun buttonAddClicked(view: View) {
        mCatView.setText("Uploading...")
        mName = getStr(name)
        //mName =( name.getText().toString() == null ? "" : name.getText().toString());
        mEmail = getStr(email)
        mPhone = getStr(phone)
        mPosition = getStr(position)
        mPassword = getStr(password)

        if (mPhone.startsWith("0")) {
            mPhone = "+84" + mPhone.substring(1)
        }

        val map: HashMap<String, Any>?
        if (putToHashMap() == null) {
            return
        } else {
            map = putToHashMap()
        }

        modifyUser(map!!, signal)

        if (signal == "updateUser") {
            uploadImage()
        }
        mCatView.dismiss()
        finish()
    }

    private fun putToHashMap(): HashMap<String, Any>? {
        val map = HashMap<String, Any>()
        if (mName == null || mName!!.isEmpty()) {
            return null
        } else {
            map["name"] = mName!!
        }

        if (mEmail == null || mEmail!!.isEmpty()) {
            return null
        } else {
            map["email"] = mEmail!!
        }

        map["phone"] = mPhone
        map["position"] = mPosition

        if (mPassword == null || mPassword!!.isEmpty()) {
            return null
        } else {
            map["password"] = mPassword!!
        }
        if (staff != null) {
            map["uidFriend"] = staff!!.uid!!
            map["avatarUrl"] = staff!!.avatarUrl!!
        } else {
            map["avatarUrl"] = "https://firebasestorage.googleapis.com/v0/b/quan-ly-nhan-vien.appspot.com/o/585e4bf3cb11b227491c339a.png?alt=media&token=01b70fd2-2eb0-4afb-a7e0-864be1a874d9"
        }
        return map
    }

    private fun openImage() {
        val intent = Intent()
        intent.type = "image/*"
        intent.action = Intent.ACTION_GET_CONTENT
        startActivityForResult(intent, IMAGE_REQUEST)
    }

    private fun uploadImage() {
        if (uriImage != null) {

            uploadTask = mStorageReference.putFile(uriImage!!)
            uploadTask!!.continueWithTask { task ->
                if (!task.isSuccessful) {
                    throw task.exception!!
                }
                mStorageReference.downloadUrl
            }.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    val downloadUri = task.result
                    if (downloadUri != null) {
                        val mUri = downloadUri.toString()
                        mDatabaseReference = FirebaseDatabase.getInstance().reference.child("users/" + staff!!.uid!!)
                        val map = HashMap<String, Any>()
                        map["avatarUrl"] = mUri
                        mDatabaseReference.updateChildren(map)
                    } else {
                        Toast.makeText(this@EditUserActivity, getString(R.string.can_not_upload_image), Toast.LENGTH_SHORT).show()
                    }
                }
            }.addOnFailureListener { e ->
                Toast.makeText(this@EditUserActivity, e.message, Toast.LENGTH_SHORT).show()
                Log.i(TAG, e.message!!)
            }
        } else {
            Toast.makeText(this, "No image selected", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == IMAGE_REQUEST && resultCode == Activity.RESULT_OK && data != null && data.data != null) {
            uriImage = data.data
            if (uploadTask != null && uploadTask!!.isInProgress) {
                Toast.makeText(this, "Upload in progress", Toast.LENGTH_SHORT).show()
            } else {
                Glide.with(this).load(uriImage).apply(RequestOptions.circleCropTransform()).into(avatar)
            }
        }
    }

    private fun binding() {
        name = findViewById(R.id.edit_text_name)
        email = findViewById(R.id.edit_text_email)
        phone = findViewById(R.id.edit_text_phone)
        position = findViewById(R.id.edit_text_position)
        add = findViewById(R.id.btn_add)
        cancel = findViewById(R.id.btn_cancel)
        avatar = findViewById(R.id.user_profile_photo_edit)
        password = findViewById(R.id.edit_text_password)
        inputPasswordLayout = findViewById(R.id.edit_text_password_layout)
    }

    private fun modifyUser(data: HashMap<String, Any>, signal: String?) {
        // Create the arguments to the callable function.
        Log.i(TAG, signal!!)
        Log.i(TAG, data.toString())
        mFunctions.getHttpsCallable(signal)
                .call(data)
                .continueWith { task ->
                    // This continuation runs on either success or failure, but if the task
                    // has failed then getResult() will throw an Exception which will be
                    // propagated down.
                    task.result!!.data as String?
                }
                .addOnCompleteListener { task ->
                    if (!task.isSuccessful) {
                        val e = task.exception
                        if (e != null) {
                            Log.i(TAG, e.toString())
                        }
                        if (e is FirebaseFunctionsException) {
                            val ffe = e as FirebaseFunctionsException?
                            val code = ffe!!.code
                            Log.i(TAG, code.toString())
                        }
                        Log.w(TAG, "addNumbers:onFailure", e)
                        showSnackBar()
                    } else {
                        val result = task.result
                        Toast.makeText(this@EditUserActivity, result, Toast.LENGTH_SHORT).show()
                    }
                }.addOnFailureListener { e -> Toast.makeText(this@EditUserActivity, e.message, Toast.LENGTH_SHORT).show() }
    }

    private fun showSnackBar() {
        Snackbar.make(findViewById(android.R.id.content), "An error occurred.", Snackbar.LENGTH_SHORT).show()
    }

    companion object {
        internal var mEmail: String? = ""
        internal var mPhone = ""
        internal var mPosition = ""
        internal var mPassword: String? = ""

        private const val IMAGE_REQUEST = 1
        private const val TAG = "kiemtra"
    }
}